### ISSUE TITLE
Ordinary module song ends up being unreadable

### ISSUE THREAD

**Issue by papiezak** [rerrahkr/BambooTracker#44]
I don't know how it happened or is it a real bug.
I made a little test - just 2 patterns, one was empty, two instruments (one was FM, second was an unused SSG instrument),50 Hz tick freq, few notes there and there and I saved result. Now when I try to open it I only see this nonsense:
`array::at:__n (which is 255) >= _Nm (which is 128)`

If it's any help, there is a module in question -  [kona.btm.zip](https://github.com/rerrahkr/BambooTracker/files/2713328/kona.btm.zip)

**Comment by rerrahkr** [comment]
Thanks for your report.
When analyzing the file, I noticed that SSG instrument number was set to 0xff that it was unexpected value. I attached the fixed file.
[kona_fixed.zip](https://github.com/rerrahkr/BambooTracker/files/2714888/kona_fixed.zip)

Did you do anything for SSG instrument except add it?

**Comment by papiezak** [comment]
>Did you do anything for SSG instrument except add it?

To be honest, I don't really remember. I think I was checking out waveform tab to see how does it work and sound - but in the end I unchecked it afted I played around with it enough

**Comment by rerrahkr** [comment]
I checked codes once again. Although I didn't know why it happens, in 758c350, I removed code which set a instrument number to -1 (0xFF). -1 was a flag as "this instrument is not registered with instrument list". I divided the flag and the instrument number, so I think it would not occur to set wrong instrument number.

**Comment by rerrahkr** [comment]
However it doesn't happen on my Windows10, I revert `master` and make new branch to fix this bug. Does it work without problems in the current `master`?

**Comment by papiezak** [comment]
OK now I dnon't know what's going on.
I git cloned the source code, compiled it and looks like  b04bba0  didn't fix it, it still crashes!

Maybe it's a problem with 3c4e88f, I will check...

**Comment by papiezak** [comment]
I tested current master on Debian Stable 9.7 with KDE. interestingly, now it doesn't crash, but I can't use up/down keys to lay notes (second note overwrites the first one in a step). To go to a different step, or down the pattern, I have to use a mouse. Aside from that, everything works fine

**Comment by rerrahkr** [comment]
I think you might set editable step to 0. I noticed that I forget to set default editable step size to 1. I will fix it.

**Comment by papiezak** [comment]
>I think you might set editable step to 0

Editable step was by default, after building set to 256, I didn't catch that :rofl:

**Comment by rerrahkr** [comment]
Does it work well by doing `make clean` and rebuilding all? While developing, I have met bugs fixed by rebuilding several times.

**Comment by papiezak** [comment]
I don't have that ubuntu partition anymore. But I tried out `fix-instrument-number-error` branch and after make clean, rebuild and `sudo make install` it doesn't crash as well. 
I don't know what was wrong there on Ubuntu. Maybe a random compilation error. 

I also tried to replicate the bug with  009ea77 fix, and I failed to meet any .btm corruption. Maybe an even more in-depth test is needed, but there's already #59 for that 
Anyway, if I'll have any problem with BambooTracker under linux, I will make a separate issue. Closing.
